L044, L045: Handle Exasol VALUES clause

barrywhart opened the pull request:
<!--Firstly, thanks for adding this feature! Secondly, please check the key steps against the checklist below to make your contribution easy to merge.-->

<!--Please give the Pull Request a meaningful title (including the dialect this PR is for if it is dialect specific), as this will automatically be added to the release notes, and then the Change Log.-->

### Brief summary of the change made
Fixes #2362

Enhances `SelectCrawler` class to interpret `VALUES` clause in a CTE as an equivalent `SELECT`.
<!--If there is an open issue for this, then please include `fixes #XXXX` or `closes #XXXX` replacing `XXXX` with the issue number and it will automatically close the issue when the pull request is merged. Alternatively if not fully closed you can say `makes progress on #XXXX` to create a link on that issue without closing it.-->


### Are there any other side effects of this change that we should be aware of?


### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

jpy-git commented on test/fixtures/rules/std_rule_cases/L044.yml:
```diff
@@ -514,3 +514,13 @@ test_snowflake_delete_cte:
     core:
       dialect: snowflake
+
+test_pass_exasol_values_clause_cte:
+  pass_str: |
+    WITH txt AS (
+        VALUES (1)
+        AS t (id)
+    )
+
+    SELECT *
+    FROM txt
```
This isn't valid snowflake/postgres/mysql
![image](https://user-images.githubusercontent.com/80432516/150544369-d86505b6-b378-437d-a942-164a8df90874.png)
so maybe use
```yml
configs:
    core:
      dialect: exasol
```
in these test cases

jpy-git replied:
This however is valid snowflake/postgres (although annoyingly doesn't parse in postgres atm) and also raises the same L044 issue:
- On main branch
![image](https://user-images.githubusercontent.com/80432516/150546227-0420e9f3-2867-40e4-8ecd-15c1f04344a9.png)

- On your branch:
![image](https://user-images.githubusercontent.com/80432516/150546677-e5cc37b5-8e32-4a41-bdf8-9befa2d53fbb.png)

L044 still there and catastophic failures 😱 🔥

jpy-git replied:
(I'll fix the postgres parsing for my example above since I know exactly where I've accidently introduced this, but that's not related to the above issue 😄 )

jpy-git replied:
Came up with the equivalent MySQL test case but it won't parse so I've added it into issue #2402. I have a feeling if it could parse then it would likely be solved by your fix for the postgres/snowflake example so we can leave it for a later date for now.

barrywhart replied:
Oops, I'll add the `exasol` config. I intended to do that, but forgot. Will also add the same test for `snowflake` dialect.

barrywhart replied:
I added a `postgres` test case as well.

barrywhart replied:
@jpy-git: Think we could create a separate issue for your `SELECT *` test case? I agree it should be fixed, but I think it's less likely to occur in production code. Happy to look into it, but I'd rather address it later (potentially next release).

jpy-git replied:
Hmm it's kind of the same as the existing test case just in a subquery rather than a CTE, and we'd be introducing catastrophic errors in rules L025 + L026. So don't think we should be merging this before at least the L025 + L026 [errors](https://github.com/sqlfluff/sqlfluff/pull/2400#discussion_r789718402) are fixed.

Worst case this runs into next release.

barrywhart replied:
I'll try and take a look this afternoon. I often have a couple of hours to work, then I don't always know when I'll have time again. 👶

barrywhart replied:
@jpy-git: I found a parsing issue with certain uses of `VALUES`. I created an issue in case you want to look at it: #2406. For now, I'll focus on the ones that parse correctly.

tunetheweb commented on test/fixtures/rules/std_rule_cases/L044.yml:
```diff
@@ -514,3 +514,13 @@ test_snowflake_delete_cte:
+test_pass_exasol_values_clause_cte:
+  pass_str: |
+    WITH txt AS (
+        VALUES (1)
```
Can we have an example with multiple values?

barrywhart replied:
Can do. I will say, though, the new code doesn't look at that part at all, AFAIK.

jpy-git commented on test/fixtures/rules/std_rule_cases/L044.yml:
```diff
@@ -514,3 +514,55 @@ test_snowflake_delete_cte:
+    core:
+      dialect: exasol
+
+test_pass_snowflake_values_clause_cte:
+  pass_str: |
+    WITH txt AS (
+        VALUES (1)
+        AS t (id)
+    )
+
+    SELECT *
+    FROM txt
+  configs:
+    core:
+      dialect: snowflake
+
+test_pass_postgres_values_clause_cte:
+  pass_str: |
+    WITH txt AS (
+        VALUES (1)
+        AS t (id)
+    )
+
+    SELECT *
+    FROM txt
+  configs:
+    core:
+      dialect: postgres
```
```suggestion
```
This isn't valid snowflake/postgres so we shouldn't add it
![image](https://user-images.githubusercontent.com/80432516/150563730-a096a95b-c822-4038-b7cf-34a07b35814e.png)

jpy-git commented on test/fixtures/rules/std_rule_cases/L045.yml:
```diff
@@ -172,3 +172,55 @@ test_snowflake_delete_cte:
+  configs:
+    core:
+      dialect: exasol
+
+test_pass_snowflake_values_clause_cte:
+  pass_str: |
+    WITH txt AS (
+        VALUES (1)
+        AS t (id)
+    )
+
+    SELECT *
+    FROM txt
+  configs:
+    core:
+      dialect: snowflake
+
+test_pass_postgres_values_clause_cte:
+  pass_str: |
+    WITH txt AS (
+        VALUES (1)
+        AS t (id)
+    )
+
+    SELECT *
+    FROM txt
+  configs:
+    core:
+      dialect: postgres
```
```suggestion
```
Not valid syntax

barrywhart commented:
@jpy-git, @tunetheweb: I believe I've addressed the PR feedback, including the additional fixes.

The only failure(s) I'm aware of now are caused by issue #2406 (which I created today). That's apparently a dialect issue which I think we should address separately. No other changes to `SelectCrawler` or the rules should be necessary once that's fixed; things will "just work".

jpy-git reviewed (approved):
@barrywhart Looks good, I've tested and can confirm those L025/L026 error have gone too 👍 LGTM 🚀 ✨ 

As you said I think once we get that remaining Postgres/MySQL/Exasol Values grammar parsing, they should be handled by this rule in the exact same way. I'm looking at that in #2404 so will try and add extra unit tests to that to cover these statements
